TFNetworkRecLayer.py

```diff
@@ -4460,3 +4460,283 @@ def __call__(self, inputs, state, scope=None):
+    if forward_weights_init is not None:
+      self._fwd_weights_initializer = get_initializer(
+        forward_weights_init, seed=self.network.random.randint(2**31), eval_local_ns={"layer": self})
+    if recurrent_weights_init is not None:
+      self._rec_weights_initializer = get_initializer(
+        recurrent_weights_init, seed=self.network.random.randint(2**31), eval_local_ns={"layer": self})
+    if bias_init is not None:
+      self._bias_initializer = get_initializer(
+        bias_init, seed=self.network.random.randint(2**31), eval_local_ns={"layer": self})
+    if self._rec_weights_initializer:
+      default_var_initializer = self._rec_weights_initializer
+    elif self._fwd_weights_initializer:
+      default_var_initializer = self._fwd_weights_initializer
+    else:
+      default_var_initializer = xavier_initializer(seed=self.network.random.randint(2**31))
+    with reuse_name_scope("rec-twod", initializer=default_var_initializer) as scope:
+      assert isinstance(scope, tf.VariableScope)
+      self._rec_scope = scope
+      scope_name_prefix = scope.name + "/"  # e.g. "layer1/rec/"
+      with self.var_creation_scope():
+        print("============!!!!!!!!!!!!========= Cell created")
+        self.cell = self._get_cell(unit_opts=unit_opts)
+
+      # this must not be part of var_creation_scope - otherwise the used operations appear to TF to be used outside
+      # of the while loop, leading to errors
+      y = self._get_output_native_rec_op(self.cell)
+
+      self.output.placeholder = y
+
+      # Very generic way to collect all created params.
+      # Note that for the TF RNN cells, there is no other way to do this.
+      # Also, see the usage of :func:`LayerBase.cls_layer_scope`, e.g. for initial vars.
+      params = tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope=re.escape(scope_name_prefix))
+      self._add_params(params=params, scope_name_prefix=scope_name_prefix)
+
+  @classmethod
+  def get_out_data_from_opts(cls, sources, **kwargs):
+    # This method may be called before the optional parent rec layer is initialized
+    # In this case, it doesn't have the `cell` attribute, and have_rec_step_info() would fail
+    # However, this only happens before the optimization of the RecLayer
+    # Therefore, if no `cell` attribute is found, and the parent layer is a RecLayer, this layer is definitely still
+    # inside of the RecLayer
+    parent_layer = kwargs["network"].get_rec_parent_layer()
+    has_cell_attribute = hasattr(parent_layer, "cell")
+    if (has_cell_attribute and kwargs["network"].have_rec_step_info()) or \
+      (not has_cell_attribute and isinstance(parent_layer, RecLayer)):
+      shape = (kwargs["n_out"],)
+      time_dim_axis = None
+      batch_dim_axis = 0
+      size_placeholder = {}
+    else:
+      shape = (None, kwargs["n_out"],)
+      time_dim_axis = 0
+      batch_dim_axis = 1
+      size_placeholder = sources[1].output.size_placeholder.copy()
```

Comment:
Do not add workarounds. Esp not such fragile workarounds which rely on some specific random behavior which could maybe change at some future point.
If there is a problem, the problem itself should be fixed, or if there is no good way to get some information, just introduce that. If needed. In this cases, it should not really be needed.
As I said, look just at how other layers do that, which is much simpler.
E.g. `SelfAttentionLayer`, or `RnnCellLayer`. Or `WindowLayer`, or `CumsumLayer`.
I guess the output shape is anyway very simple, just in all cases like `sources[1].output.shape`, or not?
Or you could check for `sources[1].output.time_dim_axis is None` or so.